share/spack/spack-completion.bash

```diff
@@ -0,0 +1,663 @@
+    fi
+}
+
+function _spack_unload {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_installed_packages)"
+    fi
+}
+
+function _spack_unuse {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_installed_packages)"
+    fi
+}
+
+function _spack_url-parse {
+    compgen -W "-h --help -s --spider" -- "$cur"
+}
+
+function _spack_urls {
+    compgen -W "-h --help -c --color -e --extrapolation" -- "$cur"
+}
+
+function _spack_use {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_installed_packages)" -- "$cur"
+    fi
+}
+
+function _spack_versions {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_all_packages)" -- "$cur"
+    fi
+}
+
+function _subcommands {
+    spack help | grep "^    [a-z]" | awk '{print $1}'
+}
+
+function _all_packages {
+    spack list
+}
+
+function _installed_packages {
+    # Perl one-liner used to strip out color codes
+    spack find | grep -v "^--" | perl -pe 's/\e\[?.*?[\@-~]//g'
+}
+
+function _installed_compilers {
+    spack compilers | egrep -v "^(-|=)"
+}
+
+function _mirrors {
+    spack mirror list | awk '{print $1}'
+}
+
+function _repos {
+    spack repo list | awk '{print $1}'
+}
+
+function test_vars {
+    echo "-----------------------------------------------------"            >> temp
+    echo "Full line:                '$COMP_LINE'"                           >> temp
+    echo                                                                    >> temp
+    echo "Word list w/ flags:       $(pretty_print COMP_WORDS[@])"          >> temp
+    echo "# words w/ flags:         '${#COMP_WORDS[@]}'"                    >> temp
+    echo "Cursor index w/ flags:    '$COMP_CWORD'"                          >> temp
+    echo                                                                    >> temp
+    echo "Word list w/out flags:    $(pretty_print COMP_WORDS_NO_FLAGS[@])" >> temp
+    echo "# words w/out flags:      '${#COMP_WORDS_NO_FLAGS[@]}'"           >> temp
+    echo "Cursor index w/out flags: '$COMP_CWORD_NO_FLAGS'"                 >> temp
+    echo                                                                    >> temp
+    echo "Subfunction:              '$subfunction'"                         >> temp
+    if $list_options
+    then
+        echo "List options:             'True'"  >> temp
+    else
+        echo "List options:             'False'" >> temp
+    fi
+    echo "Current word:             '$cur'"  >> temp
+    echo "Previous word:            '$prev'" >> temp
+}
+
+# Pretty-prints one or more arrays
+# Syntax: pretty_print array1[@] ...
+function pretty_print {
+    for arg in $@
+    do
+        local array=("${!arg}")
+        echo -n "$arg: ["
+        printf   "'%s'" "${array[0]}"
+        printf ", '%s'" "${array[@]:1}"
+        echo "]"
+    done
+}
```

Comment:
`test_vars` and `pretty_print` are just testing functions. I will remove them once we decide on a final solution and they are no longer needed.